### ISSUE TITLE
A very basic example

### ISSUE THREAD

**Issue by sudar** [matthijskooijman/TStreaming#1]
Would be really useful, if we can have a very simple "Hello World" type program/sketch to show how to use this library :)

**Comment by matthijskooijman** [comment]
Agreed! However, I'm drowning in other things to do atm, so hopefully I can get back to this in a month or so.

The basic usage is quite simple, though, just include TStreaming.h and start using the << operator and formatters. Also see this sketch, which uses the operators extensively: https://github.com/matthijskooijman/arduino-max

**Comment by sudar** [comment]
> I'm drowning in other things to do atm, so hopefully I can get back to this in a month or so.

Okay no problem. I could understand :)

> The basic usage is quite simple, though, just include TStreaming.h and start using the << operator and formatters. Also see this sketch, which uses the operators extensively: https://github.com/matthijskooijman/arduino-max

Thanks. Will check it out.

**Comment by sudar** [comment]
I tried out a simple sketch, but couldn't get it to work.

```
#include <TStreaming.h>

void setup() 
{
    int x = 10;

    Serial.begin(9600);
    Serial << "Value: " << x << " (0x" << V<Hex>(x) << ")";
}

void loop()
{
    ;
}
```

but I am getting the following error

`no match for ‘operator<<’ in ‘Serial << "Value: "’`

PS: I know that you are busy. So nothing urgent. Reply to this when you get some free time.

**Comment by matthijskooijman** [comment]
Weird, I'd have said this sketch should work, not sure why.

Your sketch compiles for me, with this Makefile:

```
    # Use git checkout of Arduino
    ARDUINO_DIR=/home/matthijs/docs/Electronics/Arduino/Arduino

    BOARD_TAG    = uno
    ARDUINO_PORT = /dev/ttyACM0
    #BOARD_TAG    = ethernet
    #ARDUINO_PORT = /dev/ttyUSB0

    # Default to uploading
    all: upload
            $(MAKE) upload ARDUINO_PORT=/dev/ttyACM1
    include $(ARDMK_DIR)/arduino-mk/Arduino.mk


    CXXFLAGS+=-std=c++11

    # Make sure the tty is freed before starting an upload
    reset: free_tty

    TTY_PIDS=$(shell lsof -t $(ARDUINO_PORT))

    free_tty:
            [ -z "$(TTY_PIDS)" ] || (kill -9 $(TTY_PIDS))

    .PHONY: free_tty
```

Perhaps you are missing the c++11 flag (not sure if I documented that
anywhere). I wouldn't expect it to give that error without that flag, though,
though.

I'm using an Arduino checkout from git, branch master revision a138309 (so
1.x).

**Comment by sudar** [comment]
I am also running Arduino 1.x (1.0.5)

I guess the compiler that we are using are different.

By adding `-std=c++11` I get the following error

```
/usr/bin/avr-g++ -x c++ -include Arduino.h -MMD -c -mmcu=atmega32u4 -DF_CPU=16000000L -DARDUINO=105 -I. -I/home/sudar/apps/arduino-1.0.5/hardware/arduino/cores/arduino -I/home/sudar/apps/arduino-1.0.5/hardware/arduino/variants/leonardo  -I/home/sudar/Dropbox/code/arduino-sketches/libraries/TStreaming -g -Os -Wall -ffunction-sections -fdata-sections -DUSB_VID=0x2341 -DUSB_PID=0x8036 -std=c++11 -fno-exceptions   PrintHex.ino -o build-leonardo/PrintHex.o
cc1plus: error: unrecognized command line option "-std=c++11"
make: *** [build-leonardo/PrintHex.o] Error 1
```

My avr-g++ version is

```
$ avr-g++ --version
avr-g++ (GCC) 4.5.3
```

**Comment by matthijskooijman** [comment]
Ah, your gcc version is too old. I'm using Arduino on Debian, which does not use the avr-gcc version bundled with Arduino, but uses the regular avr-gcc package which is at 4.7.2.

Since TStreaming really depends on C++11 for some templating constructs, you won't be able to use it with your avr-gcc version...

**Comment by sudar** [comment]
Actually I am using the latest version of avr-gcc package for Debian (Ubuntu). Did you install your package using `apt-get` or `aptitude` or you manually compiled it?

The avr-gcc that comes with Arduino is still old. It is `4.3.2`

**Comment by matthijskooijman** [comment]
I'm using this one, so just the default one installed with apt: http://packages.debian.org/wheezy/gcc-avr

Perhaps your machine is still on squeeze (which is oldstable since a month or so now)?

**Comment by matthijskooijman** [comment]
Did you manage to sort this out?

**Comment by sudar** [comment]
> Did you manage to sort this out?

Unfortunately no.

I tried installing the `gcc-avr` but it is only installing `4.5.3` in my Ubuntu desktop. Couldn't find a way to install `4.7.2` that is needed to compile this library.

**Comment by matthijskooijman** [comment]
The most  recent Arduino versions (1.5.8 and up, I think) have a newer toolchain included.   C++11 is still not enabled by default (see https://github.com/arduino/Arduino/pull/2175), though.

Did you ever try using this library again?

**Comment by sudar** [comment]
I was not able to get it to working earlier and left it at that.

I will give it a try again this weekend and will let you know if I was able to get it working or not.

**Comment by sudar** [comment]
@matthijskooijman I just tried the sample sketch with Arduino version 1.6.1 and it works.

PS: Thanks for reminding me. I almost forgot about the library :) Also thanks for trying to get C++11 into Arduino. It would be nice if your pull request get's merged into Arduino.
